### PR TITLE
Add MOI wrapper

### PR DESCRIPTION
**Pull request by blegat**
Closes https://github.com/JuliaSmoothOptimizers/NLPModelsJuMP.jl/issues/152

### CODE REVIEW

File: Project.toml [hunk 1]
```diff
@@ -7,19 +7,23 @@ JuMP = "4076af6c-e467-56ae-b986-b466b2749572"
 LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
 MathOptInterface = "b8f27783-ece8-5eb3-8dc8-9495eed66fee"
 NLPModels = "a4795742-8479-5a88-8948-cc11e1c8c1a6"
+NLPModelsTest = "7998695d-6960-4d3a-85c4-e1bceb8cd856"
```
Comment (tmigot) [inline Project.toml:10]: Do we need to add this in the deps?
```suggestion
```

File: src/MOI_wrapper.jl [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+    solver
+    nlp::Union{Nothing,MathOptNLPModel}
+    stats::Union{
+        Nothing,
+        SolverCore.GenericExecutionStats{Float64,Vector{Float64},Vector{Float64},Any},
+    }
+    function Optimizer()
+        return new(Dict{String,Any}(), false, nothing, nothing, nothing)
+    end
+end
```
Comment (tmigot) [inline src/MOI_wrapper.jl:15]: Wouldn't it be more efficient to have a non-mutable struct defined for a given problem? At leat for the stats, doing `reset!(stats)` re-initializes all the fields.

File: src/MOI_wrapper.jl [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+
+# FIXME return the name of the underlying NLPModel solver
+MOI.get(::Optimizer, ::MOI.SolverName) = "NLPModels"
```
Comment (tmigot) [inline src/MOI_wrapper.jl:18]: Could we access `MOI.RawOptimizerAttribute("solver")` to make a more precise name?

File: src/MOI_wrapper.jl [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+    return optimizer.stats.objective
+end
+
+function MOI.get(optimizer::Optimizer, attr::MOI.PrimalStatus)
+    if attr.result_index > MOI.get(optimizer, MOI.ResultCount())
+        return MOI.NO_SOLUTION
+    elseif MOI.get(optimizer, MOI.TerminationStatus()) == MOI.LOCALLY_SOLVED
+        return MOI.FEASIBLE_POINT
+    else
+        # TODO
```
Comment (tmigot) [inline src/MOI_wrapper.jl:164]: What needs to be done here?

**Review by tmigot — CHANGES_REQUESTED** [review]
Thanks @blegat for the PR! That's a great update. I made a couple of comments/questions, but nothing big I think.

File: src/MOI_wrapper.jl [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+        return MOI.UNKNOWN_RESULT_STATUS
+    end
+end
+
+function MOI.get(::Optimizer, ::MOI.DualStatus)
+    # TODO
```
[the remark this answers is not in the document]
Comment (blegat) [reply src/MOI_wrapper.jl:170]: Similar to `PrimalStatus` but for the dual variables. So if there is no dual available, just return `NO_SOLUTION`. If you have one and you know it's feasible, return `FEASIBLE_POINT`, ...

File: src/MOI_wrapper.jl (hunk 4 above)
Comment (blegat) [reply src/MOI_wrapper.jl:164]: You have a better understanding of what's the status of the primal given the `optimizer.stats`, ...
See here fore the list

File: src/MOI_wrapper.jl (hunk 2 above)
Comment (blegat) [reply src/MOI_wrapper.jl:15]: But then how do we update it with what is returned by `solve!`. Isn't that small overhead negligible as it's not in the hot loop ?

File: src/MOI_wrapper.jl (hunk 3 above)
Comment (blegat) [reply src/MOI_wrapper.jl:18]: Yes, is there a way to get this in the NLPModels API ?

File: src/MOI_wrapper.jl [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+
+function MOI.optimize!(model::Optimizer)
+    options = Dict{Symbol,Any}(
+        Symbol(key) => model.options[key]
+        for key in keys(model.options) if key != "solver"
+    )
+    if model.silent
+        options[:verbose] = 0
+    end
+    model.stats = SolverCore.solve!(model.solver, model.nlp; options...)
```
Comment (tmigot) [inline src/MOI_wrapper.jl:106]: About updating stats in a hot loop, you should use this:
```suggestion
    SolverCore.solve!(model.solver, model.nlp, model.stats; options...)
```

File: src/MOI_wrapper.jl (hunk 2 above)
Comment (tmigot) [reply src/MOI_wrapper.jl:15]: I made a comment on your solve! call for this

File: src/MOI_wrapper.jl (hunk 3 above)
Comment (tmigot) [reply src/MOI_wrapper.jl:18]: Do you mean to access the solver name ? We don't really have that now unfortunately, but it is a good suggestion. Something like
```
get_solver_name(solver)
```
would do the trick.

Comment (tmigot) [reply src/MOI_wrapper.jl:18]: I opened an issue for this https://github.com/JuliaSmoothOptimizers/SolverCore.jl/issues/73

File: src/MOI_wrapper.jl (hunk 4 above)
Comment (tmigot) [reply src/MOI_wrapper.jl:164]: Well, there is no really other status than `:first_order` that ensures that you have a ''feasible'' point. That might also be interesting information to add.
You can have access to `stats.primal_feas` which is the norm if primal feasibility if that helps.

If nothing can be added let's keep it for later.

File: src/MOI_wrapper.jl (hunk 5 above)
[the remark this answers is not in the document]
Comment (tmigot) [reply src/MOI_wrapper.jl:170]: Same as above. Just a complementary question, what do you mean by dual feasibility when you have bounds and inequalities ? I am asking as there could be more than one answer hère I think.

**Review by tmigot — COMMENTED** [review]
Thanks again @blegat , I just made a couple more comments but we should be very close now.

File: src/MOI_wrapper.jl [hunk 7]
```diff
@@ -103,7 +103,7 @@ function MOI.optimize!(model::Optimizer)
     if model.silent
         options[:verbose] = 0
     end
```
Comment (tmigot) [inline src/MOI_wrapper.jl:105]:
```suggestion
    else
      options[:verbose] = 1
    end
```

File: src/MOI_wrapper.jl [hunk 8]
```diff
@@ -22,7 +25,7 @@ MOI.is_empty(optimizer::Optimizer) = isnothing(optimizer.solver) && isnothing(op
 function MOI.empty!(optimizer::Optimizer)
     optimizer.solver = nothing
     optimizer.nlp = nothing
-    optimizer.stats = nothing
+    # TODO how to I reset `stats` ?
```
Comment (tmigot) [inline src/MOI_wrapper.jl:28]: There is a SolverCore.reset! function for that : 
```suggestion
    reset!(stats)
```

Comment (tmigot) [reply src/MOI_wrapper.jl:28]: https://github.com/JuliaSmoothOptimizers/SolverCore.jl/blob/e310d911ea8de145979584c9af8ada2d02e77c90/src/stats.jl#L178

Comment (tmigot) [reply src/MOI_wrapper.jl]: Ahaha! So true

**Review by tmigot — APPROVED** [review]
Thanks @blegat for initiating this new feature. The PR looks good to me now.

**Pull request merged**